// sbtc/dist/esm/constants.js
import * as btc from '@scure/btc-signer';
export const SBTC_PEG_ADDRESS_TESTNET = '--TODO--';
export const SBTC_PEG_ADDRESS_DEVENV = 'bcrt1pte5zmd7qzj4hdu45lh9mmdm0nwq3z35pwnxmzkwld6y0a8g83nnqhj6vc0';
export const SBTC_FT_ADDRESS_TESTNET = 'ST3ACFZ2XKFPKAFD9YY92ZP0Y4QN6MTX3H20C27NY.asset';
export const SBTC_FT_ADDRESS_DEVENV = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.asset';
export var MagicBytes;
(function (MagicBytes) {
    MagicBytes["Mainnet"] = "X2";
    MagicBytes["Testnet"] = "T2";
    MagicBytes["Regtest"] = "id";
})(MagicBytes || (MagicBytes = {}));
export var OpCode;
(function (OpCode) {
    OpCode["PegIn"] = "3C";
    OpCode["PegOut"] = "3E";
})(OpCode || (OpCode = {}));
export const MAINNET = {
    ...btc.NETWORK,
    bip32: {
        public: 0x0488b21e,
        private: 0x0488ade4,
    },
    bip84: {
        coin: 0,
    },
    magicBytes: MagicBytes.Mainnet,
};
export const TESTNET = {
    ...btc.TEST_NETWORK,
    bip32: {
        public: 0x043587cf,
        private: 0x04358394,
    },
    bip84: {
        coin: 1,
    },
    magicBytes: MagicBytes.Testnet,
};
export const REGTEST = {
    ...TESTNET,
    bech32: 'bcrt',
    magicBytes: MagicBytes.Regtest,
};
export const OVERHEAD_TX = 4 + 1 + 1 + 4;
export const VSIZE_INPUT_P2WPKH = 68;
